### PR TITLE
 #164798917  Users should be able edit tags and articles

### PR DESCRIPTION
**Pull request by davidokonji**
#### Type of change
- [x]  added the endpoint for editing articles
- [x] added the documentations for endpoints
#### How Has This Been Tested?
Screenshot 2019-04-12 at 13 07 18

#### Checklist:
- [x]  added endpoint for editing articles
- [x] added documentations for endpoints
#### Story (PT)
[#164798917] (https://www.pivotaltracker.com/story/show/164798917)

### CODE REVIEW

File: server/controllers/article.controller.js [hunk 1]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
     */
     if (req.file) req.body.image = req.file.secure_url;
     const {
-      title, body, tags, description, image, published
+      title, body, tagsList, description, published
     } = req.body;
     const { id } = req.user;
     const successMessage = 'Article successfully created';
     const errorMessage = 'Article was not created successfully';
+    const tags = [...new Set(tagsList)];
```
Comment (TheDrizzyWay) [inline server/controllers/article.controller.js]: Good job. I like the initiative you took in the aspect of filtering the tags. For the sake of consistency throughout the project, and to stay in line with sequelize conventions, I suggest you name your models starting with a capital letter.  So 'tag' should be 'Tag' and articletag should be 'Articletag'.

Comment (davidokonji) [reply server/controllers/article.controller.js]: Ok thank you, would follow the model conventions on going

File: server/controllers/article.controller.js [hunk 2]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
...
+      slug: slugify(`${title} ${Date.now()}`),
       userId: id
     });
+    if (tags) {
+      let insertTags = tags.map(async (newTags) => {
+        const newtag = await tag.findOrCreate({ where: { name: newTags } });
+        return newtag[0];
+      });
+
+      insertTags = await Promise.all(insertTags);
```
Comment (Chiazokam) [inline server/controllers/article.controller.js:45]: Hi @davidokonji  A little explanation on the `Promise.all` please.

File: server/migrations/20190409173057-create-tags.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        allowNull: false,
+        type: Sequelize.DATE,
+        defaultValue: Sequelize.fn('now')
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('tags');
+  }
+};
```
Comment (Chiazokam) [inline server/migrations/20190409173057-create-tags.js]: EOL. It is good practice, you know.

**Review by Chiazokam — CHANGES_REQUESTED** [review]

File: server/migrations/20190409173057-create-tags.js (hunk 3 above)
Comment (davidokonji) [reply server/migrations/20190409173057-create-tags.js]: Ok thank you would make corrections

File: server/routes/article.routes.js [hunk 4]
```diff
@@ -1,19 +1,31 @@
...
 
-router.post('/', [signInAuth, imageUpload, createArticle], tryCatch(ArticleController.createArticle));
+router.post('/', [signInAuth, imageUpload, createarticle], tryCatch(createArticle));
 
 router.post('/:slug/comment', [checkComments, signInAuth], tryCatch(createComments));
 
 router.post('/:slug/comment/:commentid/thread', [checkComments, signInAuth], tryCatch(threadedComment));
 
 router.post('/:articleId/claps', signInAuth, tryCatch(ClapController.createClap));
+router.put('/:slug', [signInAuth, checkArticleExist, editarticle], tryCatch(editArticle));
```
Comment (Chiazokam) [inline server/routes/article.routes.js]: Who can edit the article? All signed in users?

File: server/controllers/article.controller.js (hunk 2 above)
Comment (davidokonji) [reply server/controllers/article.controller.js:45]: @Chiazokam the `Promise.all` is used to resolve the promise which was sent from the `insertTags` array which returns a promise

File: server/routes/article.routes.js (hunk 4 above)
Comment (davidokonji) [reply server/routes/article.routes.js]: @Chiazokam Thank you for pointing this to me, only the author of the article should be able to edit an article

File: server/controllers/article.controller.js (hunk 2 above)
Comment (Chiazokam) [reply server/controllers/article.controller.js:45]: Alright..thanks

File: server/controllers/article.controller.js [hunk 5]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
...
+        firstname,
+        email,
+        bio,
+        image
+      }
+    };
+    if (!editarticle) {
+      return Response(res, 400, 'error editing article');
+    }
+    return Response(res, 200, 'edited successfully', response);
```
Comment (chukwudidu) [inline server/controllers/article.controller.js]: Good job. Maybe we could make the response message here more descriptive.

File: server/controllers/article.controller.js [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import db from '../models';
-import slug from '../utils/slugify.utils';
+import slugify from '../utils/slugify.utils';
 import Response from '../utils/response.util';
 
-const { Article } = db;
+const { Article, tag, articletag } = db;
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: There is inconsistency in table naming @davidokonji

File: server/controllers/article.controller.js [hunk 7]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
...
     const errorMessage = 'Article was not created successfully';
+    const tags = [...new Set(tagsList)];
     const article = await Article.create({
       title,
       body,
-      tags,
+      tagsList,
       description,
-      image,
+      image: req.body.image,
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: Why didn't you destructure `image`?

File: server/controllers/article.controller.js [hunk 8]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
...
+      await article.addTags(insertTags);
+    }
+    const {
+      dataValues: {
+        firstname,
+        email,
+        bio,
+        image
+      }
+    } = await article.getUser();
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: Good destructuring here

File: server/controllers/article.controller.js [hunk 9]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
     */
     if (req.file) req.body.image = req.file.secure_url;
     const {
-      title, body, tags, description, image, published
+      title, body, tagsList, description, published
     } = req.body;
     const { id } = req.user;
     const successMessage = 'Article successfully created';
     const errorMessage = 'Article was not created successfully';
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: No need for these variables here

File: server/controllers/article.controller.js [hunk 10]
```diff
@@ -20,25 +20,123 @@ class ArticleController {
...
+   */
+  static async editArticle(req, res) {
+    const {
+      params: { slug },
+      body: {
+        description, title, body, tagsList
+      }, user: { id }
+    } = req;
+    const tags = [...new Set(tagsList)];
+    const editarticle = await Article.update({
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: Wrong variable naming

File: server/controllers/article.controller.js (hunk 5 above)
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: What was edited successfully? @davidokonji

File: server/routes/article.routes.js [hunk 11]
```diff
@@ -1,19 +1,32 @@
...
 
-router.post('/', [signInAuth, imageUpload, createArticle], tryCatch(ArticleController.createArticle));
+router.post('/', [signInAuth, imageUpload, createarticle], tryCatch(createArticle));
 
 router.post('/:slug/comment', [checkComments, signInAuth], tryCatch(createComments));
 
 router.post('/:slug/comment/:commentid/thread', [checkComments, signInAuth], tryCatch(threadedComment));
 
 router.post('/:articleId/claps', signInAuth, tryCatch(ClapController.createClap));
+router.put('/:slug', [signInAuth, checkArticleExist, checkAuthor, editarticle], tryCatch(editArticle));
```
Comment (paradisekelechi) [inline server/routes/article.routes.js]: Wrong variable naming `editarticle`!

**Review by paradisekelechi — CHANGES_REQUESTED** [review]

File: server/controllers/article.controller.js (hunk 6 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: Yes would effect the change

File: server/controllers/article.controller.js (hunk 7 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: @paradisekelechi  I destructed initially, but the name image was conflicting with the image I destructured for the response once the user has created the article, just to avoid the conflict

File: server/controllers/article.controller.js (hunk 8 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: Thank you

File: server/controllers/article.controller.js (hunk 9 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: Ok would work on that now

File: server/controllers/article.controller.js (hunk 10 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: Would make changes

File: server/controllers/article.controller.js (hunk 5 above)
Comment (davidokonji) [reply server/controllers/article.controller.js]: Would make the response more descriptive

File: server/routes/article.routes.js (hunk 11 above)
Comment (davidokonji) [reply server/routes/article.routes.js]: Would make changes now

File: server/controllers/article.controller.js (hunk 7 above)
Comment (paradisekelechi) [reply server/controllers/article.controller.js]: Research on how to destructure and rename at the same time

Comment (davidokonji) [reply server/controllers/article.controller.js]: @paradisekelechi  Ok would do that now

**Review by geekdanitek — APPROVED** [review]

**Review by chukwudidu — APPROVED** [review]
Looks good. Good job

File: server/controllers/article.controller.js (hunk 5 above)
Comment (BukkyOmo) [reply server/controllers/article.controller.js]: Good job!

**Review by Chiazokam — APPROVED** [review]

File: server/controllers/article.controller.js [hunk 12]
```diff
@@ -20,25 +20,121 @@ class ArticleController {
     */
     if (req.file) req.body.image = req.file.secure_url;
     const {
-      title, body, tags, description, image, published
+      title, body, tagsList, image, description, published,
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: `tagList` is a better name and not `tagsList`

File: server/controllers/article.controller.js [hunk 13]
```diff
@@ -20,25 +20,121 @@ class ArticleController {
...
     } = req.body;
     const { id } = req.user;
-    const successMessage = 'Article successfully created';
-    const errorMessage = 'Article was not created successfully';
+    const tags = [...new Set(tagsList)];
     const article = await Article.create({
       title,
       body,
-      tags,
+      tagsList: tags,
```
Comment (paradisekelechi) [inline server/controllers/article.controller.js]: `tagList`

File: server/migrations/20190325190307-create-article.js [hunk 14]
```diff
@@ -32,7 +32,7 @@ module.exports = {
         type: Sequelize.STRING,
         allowNull: true
       },
-      tags: {
+      tagsList: {
```
Comment (paradisekelechi) [inline server/migrations/20190325190307-create-article.js]: tagList

File: server/migrations/20190404145044-create-comments.js [hunk 15]
```diff
@@ -37,6 +37,6 @@ module.exports = {
     });
   },
   down: (queryInterface, Sequelize) => {
-    return queryInterface.dropTable('Comment');
+    return queryInterface.dropTable('Comments');
```
Comment (paradisekelechi) [inline server/migrations/20190404145044-create-comments.js]: Why this change to `Comments`? Kindly check the codebase and verify what is appropriate

File: server/seeders/20190407220433-articles.js [hunk 16]
```diff
@@ -48,7 +48,7 @@ module.exports = {
       body: 'His name is Joshua',
       description: 'he should be fined',
       published: true,
-      tags: ['hello', 'catch']
+      tagsList: ['hello', 'catch']
```
Comment (paradisekelechi) [inline server/seeders/20190407220433-articles.js]: `tagList` is better. The `List` indicates the plurality of the tags.

**Review by paradisekelechi — CHANGES_REQUESTED** [review]
Implement these changes accordingly

File: server/migrations/20190404145044-create-comments.js (hunk 15 above)
Comment (davidokonji) [reply server/migrations/20190404145044-create-comments.js]: @paradisekelechi It is the migrations file, by following sequelize convention migrations are stated in plural.

**Comment by BukkyOmo** [comment]
Looks good.

**Review by BukkyOmo — APPROVED** [review]

File: server/migrations/20190404145044-create-comments.js (hunk 15 above)
Comment (paradisekelechi) [reply server/migrations/20190404145044-create-comments.js]: @davidokonji That's correct!

**Review by TheDrizzyWay — APPROVED** [review]
Nice initiative on creating a find article middleware.

**Review by paradisekelechi — APPROVED** [review]
